File: lib/routes/route-display.ts
/**
 * Metadatos de presentación de rutas a partir del nombre/descripción.
 * Extrae corredores, terminales aproximadas y pistas de búsqueda.
 */

export type RouteDisplayInfo = {
  baseName: string;
  /** Colonias / avenidas / zonas mencionadas en el nombre */
  corridors: string[];
  corridorLabel: string;
  terminalIda: string;
  terminalVuelta: string;
  /** Texto de búsqueda ampliado (alias, color, número) */
  searchBlob: string;
};

const BRACKET_RE = /\[([^\]]+)\]|\(([^)]+)\)/g;

export function parseRouteDisplay(route: {
  id: string;
  name: string;
  description?: string | null;
  color?: string;
  transport_type?: string;
}): RouteDisplayInfo {
  const name = route.name || route.id;
  const corridors: string[] = [];
  let m: RegExpExecArray | null;
  const re = new RegExp(BRACKET_RE.source, 'g');
  while ((m = re.exec(name)) !== null) {
    const raw = (m[1] || m[2] || '').trim();
    if (!raw) continue;
    for (const part of raw.split(/[-–—\/,|]+/)) {
      const t = part.trim();
      if (t && t.length > 1) corridors.push(t);
    }
  }

  const baseName = name.replace(/\s*[\[(][^)\]]*[)\]]\s*/g, ' ').replace(/\s+/g, ' ').trim() || name;
  const uniqueCorridors = Array.from(new Set(corridors)).slice(0, 5);

  const terminalIda =
    uniqueCorridors[0] ||
    (uniqueCorridors.length ? uniqueCorridors.join(' · ') : 'Salida / sentido ida');
  const terminalVuelta =
    uniqueCorridors.length >= 2
      ? uniqueCorridors[uniqueCorridors.length - 1]
      : uniqueCorridors[0]
        ? `Regreso · ${uniqueCorridors[0]}`
        : 'Regreso / sentido vuelta';

  const corridorLabel =
    uniqueCorridors.length > 0
      ? uniqueCorridors.join(' · ')
      : (route.description || '').trim() || 'Recorrido en Morelia';

  const searchBlob = [
    name,
    baseName,
    route.id,
    route.description || '',
    uniqueCorridors.join(' '),
    route.transport_type || '',
    // color words often in name (Morada, Roja…)
    baseName.replace(/\d+/g, ' '),
  ]
    .join(' ')
    .toLowerCase();

  return {
    baseName,
    corridors: uniqueCorridors,
    corridorLabel,
    terminalIda,
    terminalVuelta,
    searchBlob,
  };
}

/** Disponibilidad legible a partir de status. */
export function availabilityLabel(status?: string | null): {
  label: string;
  tone: 'ok' | 'warn' | 'bad';
} {
  const s = (status || 'approved').toLowerCase();
  if (s === 'approved' || s === 'published') return { label: 'Disponible', tone: 'ok' };
  if (s === 'needs_review') return { label: 'En revisión', tone: 'warn' };
  if (s === 'rejected') return { label: 'No disponible', tone: 'bad' };
  return { label: 'Publicada', tone: 'ok' };
}

File: viamoreliaapp/lib/gis/route_display.dart
/// Port de `lib/routes/route-display.ts` (web).
class RouteDisplayInfo {
  final String baseName;
  final List<String> corridors;
  final String corridorLabel;
  final String terminalIda;
  final String terminalVuelta;
  final String searchBlob;

  const RouteDisplayInfo({
    required this.baseName,
    required this.corridors,
    required this.corridorLabel,
    required this.terminalIda,
    required this.terminalVuelta,
    required this.searchBlob,
  });
}

class RouteDisplay {
  static final _bracketRe = RegExp(r'\[([^\]]+)\]|\(([^)]+)\)');

  static RouteDisplayInfo parse({
    required String id,
    required String name,
    String? description,
    String? transportType,
  }) {
    final corridors = <String>[];
    for (final m in _bracketRe.allMatches(name)) {
      final raw = (m.group(1) ?? m.group(2) ?? '').trim();
      if (raw.isEmpty) continue;
      for (final part in raw.split(RegExp(r'[-–—\/,|]+'))) {
        final t = part.trim();
        if (t.length > 1) corridors.add(t);
      }
    }

    final baseName = name
            .replaceAll(RegExp(r'\s*[\[(][^)\]]*[)\]]\s*'), ' ')
            .replaceAll(RegExp(r'\s+'), ' ')
            .trim()
            .isEmpty
        ? name
        : name
            .replaceAll(RegExp(r'\s*[\[(][^)\]]*[)\]]\s*'), ' ')
            .replaceAll(RegExp(r'\s+'), ' ')
            .trim();

    final unique = corridors.toSet().take(5).toList();
    final terminalIda = unique.isNotEmpty
        ? unique.first
        : 'Salida / sentido ida';
    final terminalVuelta = unique.length >= 2
        ? unique.last
        : unique.isNotEmpty
            ? 'Regreso · ${unique.first}'
            : 'Regreso / sentido vuelta';
    final corridorLabel = unique.isNotEmpty
        ? unique.join(' · ')
        : (description?.trim().isNotEmpty == true
            ? description!.trim()
            : 'Recorrido en Morelia');

    final searchBlob = [
      name,
      baseName,
      id,
      description ?? '',
      unique.join(' '),
      transportType ?? '',
      baseName.replaceAll(RegExp(r'\d+'), ' '),
    ].join(' ').toLowerCase();

    return RouteDisplayInfo(
      baseName: baseName,
      corridors: unique,
      corridorLabel: corridorLabel,
      terminalIda: terminalIda,
      terminalVuelta: terminalVuelta,
      searchBlob: searchBlob,
    );
  }

  static ({String label, String tone}) availabilityLabel(String? status) {
    final s = (status ?? 'approved').toLowerCase();
    if (s == 'approved' || s == 'published') {
      return (label: 'Disponible', tone: 'ok');
    }
    if (s == 'needs_review') return (label: 'En revisión', tone: 'warn');
    if (s == 'rejected') return (label: 'No disponible', tone: 'bad');
    return (label: 'Publicada', tone: 'ok');
  }
}
